File: Python code/3d_transformation.py
import math
import matplotlib.pyplot as plt

print("Enter 8 vertices for the 3D Cuboid (x y z):")
x_orig, y_orig, z_orig = [], [], []
for i in range(8):
    x, y, z = map(float, input(f"Vertex {i+1}: ").split())
    x_orig.append(x)
    y_orig.append(y)
    z_orig.append(z)

print("\n--- 3D Transformation Menu ---")
print("1. Translation")
print("2. Rotation (Z-axis)")
print("3. Scaling")
print("4. Reflection (XY Plane)")
print("5. Shearing (X-axis relative to Z)")
choice = int(input("Choice: "))

if choice == 1:
    tx = float(input("Enter tx: "))
    ty = float(input("Enter ty: "))
    tz = float(input("Enter tz: "))
    x_trans = [x + tx for x in x_orig]
    y_trans = [y + ty for y in y_orig]
    z_trans = [z + tz for z in z_orig]
elif choice == 2:
    angle = float(input("Enter angle (in degrees): "))
    rad = math.radians(angle)
    x_trans = [x * math.cos(rad) - y * math.sin(rad) for x, y in zip(x_orig, y_orig)]
    y_trans = [x * math.sin(rad) + y * math.cos(rad) for x, y in zip(x_orig, y_orig)]
    z_trans = list(z_orig)
elif choice == 3:
    sx = float(input("Enter sx: "))
    sy = float(input("Enter sy: "))
    sz = float(input("Enter sz: "))
    x_trans = [x * sx for x in x_orig]
    y_trans = [y * sy for y in y_orig]
    z_trans = [z * sz for z in z_orig]
elif choice == 4:
    x_trans = list(x_orig)
    y_trans = list(y_orig)
    z_trans = [-z for z in z_orig]
elif choice == 5:
    shx = float(input("Enter shx: "))
    x_trans = [x + shx * z for x, z in zip(x_orig, z_orig)]
    y_trans = list(y_orig)
    z_trans = list(z_orig)

edges = [
    (0, 1), (1, 2), (2, 3), (3, 0),
    (4, 5), (5, 6), (6, 7), (7, 4),
    (0, 4), (1, 5), (2, 6), (3, 7)
]

fig = plt.figure()
ax = fig.add_subplot(111, projection="3d")

for edge in edges:
    ax.plot(
        [x_orig[edge[0]], x_orig[edge[1]]],
        [y_orig[edge[0]], y_orig[edge[1]]],
        [z_orig[edge[0]], z_orig[edge[1]]],
        "b--",
    )

for edge in edges:
    ax.plot(
        [x_trans[edge[0]], x_trans[edge[1]]],
        [y_trans[edge[0]], y_trans[edge[1]]],
        [z_trans[edge[0]], z_trans[edge[1]]],
        "r-",
    )

ax.set_xlabel("X")
ax.set_ylabel("Y")
ax.set_zlabel("Z")
plt.show()

File: Cpp code/3d_transformation.cpp
#include <graphics.h>
#include <math.h>
#include <conio.h>
#include <iostream>

using namespace std;

void Draw3DPolygon(int x[], int y[], int z[], int n) {
    int px[8], py[8];
    float perspective = 0.5;

    for (int i = 0; i < n; i++) {
        px[i] = x[i] + z[i] * perspective;
        py[i] = y[i] - z[i] * perspective;
    }

    for (int i = 0; i < n; i++) {
        line(px[i], py[i], px[(i + 1) % n], py[(i + 1) % n]);
    }
}

int main() {
    int gd = DETECT, gm;
    int n = 4;
    int x[4], y[4], z[4];
    int nx[4], ny[4], nz[4];
    int choice;

    for (int i = 0; i < n; i++) {
        cout << "Enter point (x" << i + 1 << ", y" << i + 1 << ", z" << i + 1 << "): ";
        cin >> x[i] >> y[i] >> z[i];
    }

    cout << "\n--- Enter 3D Transformation menu: ---" << endl;
    cout << "1. Translation" << endl;
    cout << "2. Rotation (Z-Axis)" << endl;
    cout << "3. Scaling" << endl;
    cout << "4. Reflection (XY Plane)" << endl;
    cout << "5. Shearing" << endl;
    cout << "Enter Choice(1-5): ";
    cin >> choice;

    initgraph(&gd, &gm, "");

    setcolor(RED);
    Draw3DPolygon(x, y, z, n);

    setcolor(WHITE);

    if (choice == 1) {
        int tx, ty, tz;
        cout << "Enter Translation (tx, ty, tz): ";
        cin >> tx >> ty >> tz;

        for (int i = 0; i < n; i++) {
            nx[i] = x[i] + tx;
            ny[i] = y[i] + ty;
            nz[i] = z[i] + tz;
        }
        Draw3DPolygon(nx, ny, nz, n);
    }
    else if (choice == 2) {
        float angle, rad;
        cout << "Enter Rotation angle around Z-axis: ";
        cin >> angle;

        rad = angle * 3.1416 / 180.0;

        for (int i = 0; i < n; i++) {
            nx[i] = x[i] * cos(rad) - y[i] * sin(rad);
            ny[i] = x[i] * sin(rad) + y[i] * cos(rad);
            nz[i] = z[i];
        }
        Draw3DPolygon(nx, ny, nz, n);
    }
    else if (choice == 3) {
        float sx, sy, sz;
        cout << "Enter Scaling factors (sx, sy, sz): ";
        cin >> sx >> sy >> sz;

        for (int i = 0; i < n; i++) {
            nx[i] = x[i] * sx;
            ny[i] = y[i] * sy;
            nz[i] = z[i] * sz;
        }
        Draw3DPolygon(nx, ny, nz, n);
    }
    else if (choice == 4) {
        for (int i = 0; i < n; i++) {
            nx[i] = x[i];
            ny[i] = y[i];
            nz[i] = -z[i];
        }
        Draw3DPolygon(nx, ny, nz, n);
    }
    else if (choice == 5) {
        float shx, shy;
        cout << "Enter Shearing factors relative to Z (shx, shy): ";
        cin >> shx >> shy;

        for (int i = 0; i < n; i++) {
            nx[i] = x[i] + shx * z[i];
            ny[i] = y[i] + shy * z[i];
            nz[i] = z[i];
        }
        Draw3DPolygon(nx, ny, nz, n);
    }

    getch();
    closegraph();
    return 0;
}
